stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb

```diff
@@ -556,11 +556,15 @@ public struct Logging${Kind}<Base: ${Kind}Type> : ${Kind}Type, LoggingType {
     return base._copyToNativeArrayBuffer()
   }
 
-  /// Copy a Sequence into an array.
-  public func _initializeTo(ptr: UnsafeMutablePointer<Base.Generator.Element>)
-    -> UnsafeMutablePointer<Base.Generator.Element> {
+  /// Copy `self` into uninitialized contiguous memory of size `capacity`,
+  /// returning one past the last element initialized, or a `nil` pointer
+  /// if the copy cannot be safely made given `capacity`.
+  public func _initializeTo(
```

Comment:
>>  `capacity` is the number of uninitialized elements that the caller
 made available in the buffer that `ptr` points to; indeed, this should
 be spelled out in the doc comment.

> It doesn't read well at the call site because it isn't clear what it's
the capacity of.  Anywhere you are going to pass an
`UnsafeMutablePointer` + length of its potential storage, it might be
better to use an `UnsafeMutableBufferPointer`... although that was not
really designed to be used with uninitialized memory, you can always get
the baseAddress.

The thing about that is that this would be an irregular use of `UnsafeMutableBufferPointer`: the implementation may *only* call `baseAddress` and `count`, as calling anything else would use uninitialized memory. This would deserve a warning in the docs, and we'd still be in a state where the call site looks confusing/wrong unless you read the documentation.

What if we replaced count with an end pointer? E.g. 
```
let end = copyContents(to: ptr, limit: ptr + capacity)
```
(`ForwardIndex.advanced(by:limit:)` uses a `limit` label in a similar context.)

>>  ```
>> /// Copy at most `capacity` elements starting from the current element into
>> /// the buffer pointed to by `ptr`, advancing iteration state and the pointer itself.
>> /// - Returns: `true` if there are more elements remaining in the sequence.
>> mutating func _copyContents(to ptr: inout UnsafeMutablePointer<Element>, withCapacity capacity: Int) -> Bool
>>  ```

> Why do you think this belongs in `IteratorProtocol`?

If we want to be able to copy elements of a sequence into a series of segments, an iterator would be the natural way to represent the progress of copying. So the copy method (wherever it is defined) would need an inout iterator to record progress, beside the parameter(s) that describe the buffer to use (i.e., a base address + length, in some form). It needs to return the number of elements copied, and a flag that indicates whether or not it reached the end of the sequence.

Putting the copy method inside `IteratorProtocol` would get rid of the inout iterator. Also, iterators already contain enough knowledge about the sequence to retrieve a single element, so it doesn't seem too far-fetched that they would also be able to copy multiple elements at once.

>> (I don't like the `inout` parameter or the boolean result though.)

> Me neither.  Why do it that way?

I am yet to find a variant that I'm entirely happy with. Perhaps this would be better:
```
mutating func _copyContents(to bufferPointer: UMP, limit: UMP) -> (end: UMP, finished: Bool)
```
My difficulty with coming up with a really nice signature is the reason I haven't followed up on this idea yet. It may indicate that the proposed concept is just too complex, but I don't think I mastered the naming conventions to a level where I could say for sure.

>> However, having `_copyContents` in `IteratorProtocol` would probably
 mean we'd need to change `Array` to use a specialized iterator other
 than `IndexingIterator`.

> ...because?

Because I don't see how we could implement `Array`'s `_copyContents` in `IndexingIterator`.

Unless we introduced an extra hook in `Collection` that `IndexingIterator` could call into? Hm. Since that hook would be in `Collection` rather than `Sequence`, it would be quite a bit simpler… This could be interesting.

> IIUC the main point of this PR is that using iterators is better than
indexing; if that is separable from the `_initializeTo` change, I'd much
rather accept that part alone, and consider other changes in a different
PR.

Wholeheartedly agreed! I'll remove the `_initializeTo` stuff from this PR and reintroduce it after a little more time in the oven.